Refactor actor blocking with unified API

obenland opened the pull request:
Fixes #2095

## Proposed changes:
* Add `Actors::normalize_identifier()` method for centralizing actor identifier normalization logic
* Simplify `Moderation::is_actor_blocked()` to accept pre-normalized URIs with consistent int parameter
* Refactor Following and Blocked Actors table classes to use unified normalization approach
* Improve separation of concerns between actor normalization and blocking logic

### Other information:

- [ ] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Go to 'Users > Following' or 'Settings > ActivityPub > Following' 
* Try following an account using different formats:
  - Webfinger format: `@user@example.com`
  - Email format: `user@example.com` 
  - URL format: `https://example.com/@user`
  - URL without scheme: `example.com/@user`
* Verify that blocked actors are properly prevented from being followed
* Go to 'Users > Blocked Actors' or 'Settings > ActivityPub > Blocked Actors'
* Try blocking accounts using the same various formats
* Verify all formats work correctly and error handling is appropriate

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch

#### Type

-   [x] Changed - for changes in existing functionality

#### Message

Refactor actor blocking with unified API for better maintainability

</details>

obenland committed:
Refactor actor blocking with unified API

- Add Actors::normalize_identifier() for consistent actor URI normalization
- Simplify Moderation::is_actor_blocked() to accept normalized URIs
- Refactor Following and Blocked Actors tables to use centralized logic
- Change is_actor_blocked() default user_id from null to 0 for type consistency
- Improve separation of concerns between normalization and blocking logic

Addresses PR review feedback requesting unified blocking API.

matticbot committed:
Add changelog

obenland committed:
Fix error message to display original profile identifier

Replaces the use of the normalized profile identifier with the original input in the error message when blocking an actor fails. This ensures the user sees the exact value they attempted to block.

obenland committed:
Merge branch 'refactor/unified-actor-blocking-check' of https://github.com/Automattic/wordpress-activitypub into refactor/unified-actor-blocking-check

obenland committed:
Update method naming to follow WordPress conventions

pfefferle commented on includes/class-moderation.php:
```diff
@@ -271,6 +271,35 @@ public static function get_site_blocks() {
+	 * @param int    $user_id   Optional. User ID to check user blocks for. Defaults to 0 (site-wide only).
+	 * @return bool True if blocked, false otherwise.
+	 */
+	public static function actor_is_blocked( $actor_uri, $user_id = 0 ) {
```
This sounds more like a statement than a question. What about

```suggestion
	public static function is_actor_blocked( $actor_uri, $user_id = 0 ) {
```

obenland replied:
For it to read like a statement is on purpose, the idea is for if statements to be read like a sentence. That said, it looks like Core prioritized consistency around the `is_` prefix over if statement sentences (`is_post_type_hierarchical()`), so I'm happy to revert back to `is_actor_blocked()`.

obenland replied:
Fixed in [1e6a1ee](https://github.com/Automattic/wordpress-activitypub/pull/2097/commits/1e6a1eedf1bd04bebbaa39ec8035d7850b6afc35).

pfefferle commented on includes/wp-admin/table/class-following.php:
```diff
@@ -123,41 +123,28 @@ public function process_action() {
-				// Check if user has blocked the account.
-				if ( \in_array( $profile, Moderation::get_user_blocks( $this->user_id )['actors'], true ) ) {
+				$original = \sanitize_text_field( \wp_unslash( $_REQUEST['activitypub-profile'] ) );
+				$profile  = Actors::normalize_identifier( $original );
```
I would move the normalizer into the `is_actor_blocked` function. It reduces the preconditions for the URL you want to check and simplifies the code.

obenland replied:
It includes prepending `https` for host-only queries, so the table functionality depends on it, too.

pfefferle replied:
I am not sure I understand?

pfefferle replied:
How does it speak against calling it directly inside of the blocked check?

obenland replied:
If it was moved into `is_actor_blocked` we'd break the functionality that allows users to put in a host only and still follow the blog actor. Like `activitypub.blog`.

The normalized identifier is a service to both `is_actor_blocked()` and `follow()` and prevents a webfinger from having to be resolved twice.

pfefferle replied:
FYI: It will not be resolved twice, we cache the request. so yes, it might be a bit of overload, but might simplify the code. but I am fine with improving over time.

obenland replied:
Dang, that's impressive. I'm looking forward to having the same detailed grasp of the plugin as you do, I did not know that off the top of my head

obenland committed:
Revert "Update method naming to follow WordPress conventions"

This reverts commit e8fc9a8aa1bfee546ed6f2c6d5e8671e00188098.

obenland committed:
Refactor normalize_identifier to use object_to_uri first

Simplifies the logic by ensuring we always work with a string after running object_to_uri() first, eliminating the need for separate object/string handling paths.

obenland committed:
Improve actor identifier normalization and add tests

Added type check to Actors::normalize_identifier to handle non-string inputs gracefully. Introduced comprehensive unit tests for normalization logic and edge cases in blocked actors, followers, and following tables, ensuring robust handling of various identifier formats and invalid inputs. Also improved test cleanup by removing filters after each test.

obenland committed:
Add HTTP request mocking for actor data in tests

Introduced a mock for HTTP requests to fetch actor data in both Test_Followers and Test_Following test classes. This ensures tests do not make real HTTP requests and improves test reliability and isolation.

obenland committed:
Reorder

obenland committed:
Clean up filter removal in blocked actors test

Moved removal of 'pre_get_remote_metadata_by_actor' filter to tear_down() and removed redundant filter removals from the test method to ensure proper test cleanup.

pfefferle commented on includes/wp-admin/table/class-following.php:
```diff
@@ -123,41 +123,28 @@ public function process_action() {
-				// Check if site has blocked the account.
-				if ( \in_array( $profile, Moderation::get_site_blocks()['actors'], true ) ) {
+				// Check if actor is blocked.
+				if ( Moderation::is_actor_blocked( $profile, $this->user_id ) ) {
```
Do we also have to check the Actor against the blocked domains?

obenland replied:
Hm, good question. Yeah probably. Maybe? What do you think?

pfefferle replied:
I would say we should.

If you block a domain, it does not make sense to follow someone from that host, because you would not receive any Activities from them.

pfefferle replied:
Should we encapsulate this in a specific function? I think I need something similar for the Starter Kits import.

pfefferle replied:
We can directly add it to `is_actor_blocked` because it seems that `check_activity_against_blocks` does not rely on that function!

obenland replied:
That makes sense, yes

obenland replied:
Fixed in [d19866b](https://github.com/Automattic/wordpress-activitypub/pull/2097/commits/d19866b3bf1b3297bca58abb857f3572dbfc61cb).

obenland committed:
Add domain checking to is_actor_blocked method and comprehensive unit tests

- Enhanced is_actor_blocked method to check both actor URIs and domains for site-wide and user-specific blocks
- Added comprehensive unit tests covering:
  - Site-wide actor and domain blocking
  - User-specific actor and domain blocking
  - Hierarchical priority (site-wide takes precedence)
  - Mixed blocking scenarios
  - Edge cases and URL format handling
  - Performance testing with many blocks
- Updated test mocks to properly support actor blocking functionality
- Improved test cleanup to handle actor post metadata

obenland committed:
Merge branch 'trunk' into refactor/unified-actor-blocking-check

obenland committed:
Merge branch 'trunk' into refactor/unified-actor-blocking-check
